### ISSUE TITLE
mut_mut suggests invalid code

### ISSUE THREAD

**Issue by jo-so** [rust-lang/rust-clippy#9035]
### Summary

Clippy suggests to re-borrow a struct field which leads to invalid code. The field is of type `&mut dyn …`

### Lint Name

mut_mut

### Reproducer

I tried this code:

```rust
#![allow(dead_code)]
#![warn(clippy::pedantic)]

use std::fmt::Display;

struct Foo<'a> {
    inner: &'a mut dyn Display,
}

impl Foo<'_> {
    fn foo(&mut self) {
        let hlp = &mut self.inner;
        bar(hlp);
    }
}

fn bar(_: &mut impl Display) { }
```

I saw this happen:

```
warning: this expression mutably borrows a mutable reference. Consider reborrowing
  --> src/main.rs:12:19
   |
12 |         let hlp = &mut self.inner;
   |                   ^^^^^^^^^^^^^^^
```

I expected to see this happen:

The reborrowing `let hlp = &mut *self.inner;` doesn't compile:

```
error[E0277]: the size for values of type `dyn std::fmt::Display` cannot be known at compilation time
  --> src/main.rs:13:13
   |
13 |         bar(hlp);
   |         --- ^^^ doesn't have a size known at compile-time
   |         |
   |         required by a bound introduced by this call
   |
   = help: the trait `Sized` is not implemented for `dyn std::fmt::Display`
```

### Version

```text
rustc 1.63.0-nightly (21e9336fe 2022-06-18)
binary: rustc
commit-hash: 21e9336fe81a1fce364349bb7a35a0347c369f34
commit-date: 2022-06-18
host: x86_64-unknown-linux-gnu
release: 1.63.0-nightly
LLVM version: 14.0.5
```
